### PR TITLE
[teams] Fix Team Plan seat de/reactivation logic

### PR DESCRIPTION
**Pull request by jankeromnes**
Attempts to fix https://github.com/gitpod-io/gitpod/issues/4693

---

I believe our current implementation has the following problems:

- Deactivating a Team Plan Seat sets a `cancellationDate` on the `d_b_team_subscription_slot` entry and its related `d_b_subscription` entry (**good**). However, it schedules the Team Plan to have 1 fewer Seat than it currently has, effective at the end of term (**bad** -- this doesn't take into account multiple deactivations in the same period):

https://github.com/gitpod-io/gitpod/blob/8552d9e67cb888dea41f2389e44ce760a87c699e/components/server/ee/src/workspace/gitpod-server-impl.ts#L1238-L1242

- Reactivating a Team Plan Seat **also decreases** the amount of Seats (by maximum 1 each billing cycle). This is obviously wrong (as it should increase the quantity instead):

https://github.com/gitpod-io/gitpod/blob/8552d9e67cb888dea41f2389e44ce760a87c699e/components/server/ee/src/workspace/gitpod-server-impl.ts#L1254-L1258

- Our "Manage Team" modal shows the correct number of active Team Plan Seats, because if filters out Seats that are now `cancelled` (1). By contrast, the total number of Members shown for your Team Plan doesn't filter out `cancelled` Seats (2), thus reporting a too-high number of "Members" (actually "Seats"):

(1)

https://github.com/gitpod-io/gitpod/blob/97b938150a2dc6211ea8ae549f08050bf1e8159f/components/dashboard/src/settings/Teams.tsx#L717-L718

(2)

https://github.com/gitpod-io/gitpod/blob/97b938150a2dc6211ea8ae549f08050bf1e8159f/components/dashboard/src/settings/Teams.tsx#L503-L504

---

This Pull Request implements the following fixes:

- When deactivating a Seat, decrease the Team Plan Seat quantity immediately, not at the end of term. I believe this allows multiple deactivations in the same cycle, without any side effects on billing.
- When reactivating a Seat, **increase** (not decrease) the Team Plan Seat quantity immediately as well (similar effect).
- Filter out cancelled Seats in the Members total count.

### How to test

Purchase a Team Plan. You should still be able to:
- Assign & unassign Seats (without impact to billing)
- Purchase additional Seats (should add a new deferred charge in Chargebee and also increase your next bill)
- Deactivate Seats (they should remain visible until the end of term, and should also decrease your upcoming bill in Chargebee)
- Deactivate multiple Seats in the same cycle (should decrease your next bill by the correct number of Seats, not just 1)
- Reactivate not-yet-cancelled Seats (the Seat should be usable again + you should not be charged again for this already-paid Seat + your next bill should have one more Seat on it)

- [x] /werft with-payment
- [ ] /werft with-clean-slate-deployment

### CODE REVIEW

**Comment by geropl** [comment]
Reviewing now.

File: components/dashboard/src/settings/Teams.tsx [hunk 1]
```diff
@@ -500,7 +500,7 @@ function AllTeams() {
                             <span className="text-sm my-auto text-gray-400 truncate overflow-ellipsis">Purchased on {formatDate(sub?.startDate)}</span>
                         </div>
                         <div className="p-0 my-auto flex flex-col w-2/12">
-                            <span className="my-auto truncate text-gray-500 overflow-ellipsis">{slots.filter(s => s.teamSubscription.id === sub.id).length || "–"}</span>
+                            <span className="my-auto truncate text-gray-500 overflow-ellipsis">{slots.filter(s => s.state !== 'cancelled' && s.teamSubscription.id === sub.id).length || "–"}</span>
```
Comment (geropl) [inline components/dashboard/src/settings/Teams.tsx:503]: :+1:

**Comment by geropl** [comment]
The frontend changes look good and make sense to me.

On the backend, the current change would improve things, allthough not completely solve it:

Good case:
 1. the user deactivates a slot
 2. Chargebee is updated
 3. we receive an Chargebee event _immediately_, and update `quantity`
 4. the user deactives another slot shortly after: :heavy_check_mark:

Bad case:
 1. the user deactivates a slot
 2. Chargebee is updated
 3. we receive an Chargebee event after _some minutes_ (happens quite often), and update `quantity`
 4. the user deactives another slot shortly after, based on the old `quantity`: :x: 
 
This is basically the same problem as before, just more unlikely. Unfortunately I don't see any other way to solve this. The alternative I thought about:
With the suggested change `end of term: false`, we could try to not use `quantity` at all, but fetch the current number from the API: This still has the conceptual TOCTOU problem, as Chargebee does not allow to atomically check `resource_version` on updates, for instance, but is another step to minimize the gap.

**Comment by geropl** [comment]
@jankeromnes AFter having slept over it the explanation is weird. Would you have some time this morning to quickly discuss in sync?

**Comment by geropl** [comment]
Verified my suspicion: The number of seats is _already_ updated immediately*: Shortly after the "change # of seats" action in Gitpod, we get back the Chargebee event ([example for `ed2aac6c-0635-4ca8-96a2-fb38a8c638cf`](https://gitpod.chargebee.com/events/ev_AzZdEUScQmMXT2RVT/details)).

So the lost update still happens, but only if a user clicks faster than Chargebee sends us updates.
Possible ways to improve this:
 1. directly query quantity from Chargebee API:
   - get rid of `ts.quantity` in our DB
   - request quantity from Chargebee API before sending a request with a manipulated quantity

   Pros:
    - reduces likelihood of lost updates
   Cons:
    - lost updates might still happen as the Chargebee API does not allow to conditionally-update a specific version of an object (like "update only if `resource_version` is equal)

 2. list slots that are "active" instead of using `ts.quantity`
   - request list of quantity from active slots
   - compare that to `ts.quantity` and issue a warning if there is a discrepancy
   
   Pros:
   - we break the cycle gitpod -> CB -> gitpod
   Cons:
   - not sure there is any

**Comment by jankeromnes** [comment]
Post-sync summary:
- It makes sense to tell Chargebee to apply these changes immediately, not at the end of term
- Using `ts.quantity` is a bad idea (because it only changes at the end of term, or after a Chargebee callback if we make the effect "immediate", but there is still some gap in which we might loose updates)
- Instead of using `ts.quantity`, it seems safer to take the base value of activations/deactivations directly from our DB (i.e. don't use whatever Chargebee thinks is correct as a base, but use whatever we show users in their Manage Team list as a base -- that's the best source of truth we have)

Will update the PR accordingly.

**Comment by jankeromnes** [comment]
I think this change is correct, and it builds successfully. Unfortunately, I can't test it:
- Hooking up chargebee-test with my `core-dev` environment doesn't work (maybe payments are disabled?)
- Trying to deploy a `io-dev` environment (by creating a new branch under gitpod-com) doesn't work ([all I get](https://werft.gitpod-io-dev.com/job/gitpod-build-jx-fix-teams.2) is "prepare" and "job config" and then Werft calls it done 🤔)

@geropl and/or @AlexTugarev, could you please review, test, and merge this change if it fixes the problem?

### How to test

1. Purchase a Team Plan with 10 seats
2. Deactivate 3 seats in quick succession
3. Teams page should still say "10 Members" (a month later it should say "7 Members" -- but that's tricky to test so don't bother)
4. Chargebee itself should say that the Team Plan has 7 Seats (and the next bill should be 7 x (price of seat), not 10x or 9x)

File: components/server/ee/src/workspace/gitpod-server-impl.ts [hunk 2]
```diff
@@ -1120,6 +1120,12 @@ export class GitpodServerEEImpl extends GitpodServerImpl<GitpodClient, GitpodSer
         return slots[0];
     }
 
+    // Get the current number of "active" slots in a team subscription (count all "assigned" and "unassigned", but not "deactivated" or "cancelled").
+    protected async tsGetActiveSlotQuantity(teamSubscriptionId: string): Promise<number> {
+        const slots = await this.teamSubscriptionDB.findSlotsByTeamSubscriptionId(teamSubscriptionId);
+        return slots.filter(slot => !!slot.cancellationDate).length;
```
Comment (AlexTugarev) [inline components/server/ee/src/workspace/gitpod-server-impl.ts:1126]: @jankeromnes just to make sure I understand this correctly...
* "active" means **no cancellation date** is set
* this filter returns slots that actually have the cancellation date set

Comment (jankeromnes) [reply components/server/ee/src/workspace/gitpod-server-impl.ts:1126]: Oh, indeed 🤦 excellent catch, thanks!!

File: components/server/ee/src/workspace/gitpod-server-impl.ts [hunk 3]
```diff
@@ -1120,6 +1120,12 @@ export class GitpodServerEEImpl extends GitpodServerImpl<GitpodClient, GitpodSer
         return slots[0];
     }
 
+    // Get the current number of "active" slots in a team subscription (count all "assigned" and "unassigned", but not "deactivated" or "cancelled").
+    protected async tsGetActiveSlotQuantity(teamSubscriptionId: string): Promise<number> {
+        const slots = await this.teamSubscriptionDB.findSlotsByTeamSubscriptionId(teamSubscriptionId);
+        return slots.filter(slot => !slot.cancellationDate).length;
```
Comment (geropl) [inline components/server/ee/src/workspace/gitpod-server-impl.ts:1126]: Would love to see an existing method re-used for this.

**Comment by AlexTugarev** [comment]
@geropl, I think we are missing `https://jx-fix-teams.staging.gitpod-dev.com/` for callbacks in the CB, because events are missing in server.

**Comment by geropl** [comment]
Tested, this seems to fix it:

## UI
 - [x] correct # of seats (now: 10)
image


## Chargbee subscription:
 - [x] correct # of seats (billed for next cycle: 5)
image

## Chargebee event:
 - [x] is sent immediately
 - [x] contains correct # of seats (billed for next cycle: 5)
image

## DB:
- [x] correct nr of seats (5 and 10, resp.)
image

**Review by geropl — APPROVED** [review]
:+1: (details see above)

**Review by geropl — CHANGES_REQUESTED** [review]
@AlexTugarev found that when you're fast enough, you can triger races in `server` between DB and multiple "deactivate slots" requests.

This is a more fundamental issue with our current approach. In the scope of this PR we'll try to:
 - introduce a queue to avoid this at least at the scope of a single server instance
 - think about a regular task that pushes the current state of the DB to Chargebee

**Review by AlexTugarev — APPROVED** [review]
Tested latest changes as described in https://github.com/gitpod-io/gitpod/pull/4713#issuecomment-875800173

It worked properly. 👍🏻

**Review by geropl — APPROVED** [review]
Fixed with https://github.com/gitpod-io/gitpod/pull/4713/commits/01e7ec09573b6c6a68d33f140a85efa896e7ef71 :tada:

**Pull request merged**